### DISCUSSION TITLE
Contributed webview context menus

### DISCUSSION THREAD

**Discussion by isidorn** [microsoft/vscode-discussions#160] [Announcements]
The new webview context menu API makes it easy for advanced [webviews](https://code.visualstudio.com/api/extension-guides/webview) to customize the context menus that shows when a user right-clicks inside a webview. This new contribution point uses VS Code's normal context menus, so custom menus fit right in with the rest of the editor. Webviews can also show custom context menus for different sections of the webview.

To add a new context menu item to your webview, first add a new entry in `menus` under the new `webview/context` section. Each contribution takes a `command` (which is also where the item's title comes from) and a `when` clause. The [when clause](https://code.visualstudio.com/api/references/when-clause-contexts) should include `webviewId == 'YOUR_WEBVIEW_VIEW_TYPE'` to make sure the context menus only apply to your extension's webviews:

```json
"contributes": {
  "menus": {
    "webview/context": [
      {
        "command": "catCoding.yarn",
        "when": "webviewId == 'catCoding'"
      },
      {
        "command": "catCoding.insertLion",
        "when": "webviewId == 'catCoding' && webviewSection == 'editor'"
      }
    ]
  },
  "commands": [
    {
      "command": "catCoding.yarn",
      "title": "Yarn 🧶",
      "category": "Cat Coding"
    },
    {
      "command": "catCoding.insertLion",
      "title": "Insert 🦁",
      "category": "Cat Coding"
    },
    ...
  ]
}
```

Inside of the webview, you can also set the contexts for specific areas of the HTML using the `data-vscode-context` [data attribute](https://developer.mozilla.org/docs/Learn/HTML/Howto/Use_data_attributes) (or in JavaScript with `dataset.vscodeContext`). The `data-vscode-context` value is a JSON object that specifies the contexts to set when the user right-clicks on the element. The final context is determined by going from the document root to the element that was clicked.

Consider this HTML for example:

```html
<div class="main" data-vscode-context='{"webviewSection": "main", "mouseCount": 4}'>
  <h1>Cat Coding</h1>

  <textarea data-vscode-context='{"webviewSection": "editor", "preventDefaultContextMenuItems": true}'></textarea>
</div>
```

If the users right-clicks on the `textarea`, the following contexts will be set:

* `webviewSection == 'editor'` — This overrides `webviewSection` from the parent element.
* `mouseCount == 4` — This is inherited from the parent element.
* `preventDefaultContextMenuItems == true` — This is a special context that hides the copy and paste entries that VS Code normally adds to webview context menus.

If the user right-clicks inside of the `<textarea>`, they will see:

![image](https://user-images.githubusercontent.com/1926584/194512818-fe5f7ae1-4f9b-47a5-9b79-7d35392575ba.png)

Try it out and let us know what you think. fyi @mjbvz

**Comment by usernamehw** [comment]
- Do webviews support nested context menus (like normal menus)?
- Do webviews support calling context menus programmatically (like <kbd>Shift</kbd>+<kbd>F10</kbd>)
